spring-http-interface: introduce `springHttpClientAdapter`, fix `paramDoc.mustache`

ibaranga opened the pull request:
Addressing issue https://github.com/OpenAPITools/openapi-generator/issues/19712 

### What
- Introduce `springHttpClientAdapter` for `spring-http-interface`
  - This property is used for selecting HTTP client implementation in Spring HTTP interfaces, with separate templates for each client configuration
- Added an `spring-http-interface`-specific **empty** `paramDoc.mustache` 
 
### Why
- Enable selecting different HTTP client implementations when generating Spring HTTP client interfaces
- Provides additional flexibility for users who want to generate non-reactive Spring Boot applications

### How
- `springHttpClientAdapter`: Allows users to choose between different HTTP client implementations used in `HttpInterfacesAbstractConfigurator`:
  - `web-client` (set by default, to ensure **backward compatibility**)
  - `rest-client`
  - `rest-template`
- Separate templates for each `HttpInterfacesAbstractConfigurator` implementation:
  - `httpInterfacesRestClientConfiguration.mustache`
  - `httpInterfacesRestTemplateConfiguration.mustache`
  - `httpInterfacesWebClientConfiguration.mustache`
- Log warning for configuration mismatch
  - When `reactive: false` is used in combination with the reactive `web-client`, it warns users of potential configuration mismatches, and suggests switching to `rest-template` or `rest-client` for non-reactive configurations.
- Remove unnecessary paramDoc
  - Added an `spring-http-interface`-specific **empty** `paramDoc.mustache` in `JavaSpring/libraries/spring-http-interface/paramDoc.mustache`
  - This prevents inheriting the `@Parameter` annotations from the default Spring template located at `JavaSpring/paramDoc.mustache`.
  - Otherwise, the generated code includes `@Parameter` annotations on request body parameters, which were causing compile errors due to missing imports

---

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@cachescrubber @welshm  @MelleD  @atextor @manedev79 @javisst @borsch @banlevente @Zomzog @martin-mfg

ibaranga commented on modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-http-interface/httpInterfacesWebClientConfiguration.mustache:
```diff
@@ -28,7 +28,7 @@ public abstract class HttpInterfacesAbstractConfigurator {
     @Bean(name = "{{configPackage}}.HttpInterfacesAbstractConfigurator.{{classVarName}}")
     {{classname}} {{classVarName}}HttpProxy() {
-        HttpServiceProxyFactory factory = HttpServiceProxyFactory.builder(WebClientAdapter.forClient(webClient)).build();
+        HttpServiceProxyFactory factory = HttpServiceProxyFactory.builderFor(WebClientAdapter.forClient(webClient)).build();
```
This is marked as deprecated, replacing it with `builderFor`, which is also in line with the rest of the `httpInterfaces*Configuration.mustache` templates

wing328 commented:
please review the errors when you've time: https://github.com/OpenAPITools/openapi-generator/actions/runs/11094295717/job/30896757555?pr=19710

wing328 commented:
please review the errors reported by the CI when you've time.

ibaranga commented:
> please review the errors reported by the CI when you've time.

Thanks again, @wing328, I've just updated the PR!

wing328 commented:
filed https://github.com/OpenAPITools/openapi-generator/pull/20129 to trigger all the tests

let's see how that goes

wing328 commented:
as shown in https://github.com/OpenAPITools/openapi-generator/pull/20129, samples are not up-to-date

can you please fix it when you've time?

wing328 commented:
@davidkarlsen do you mind doing a test locally for that?

```
git checkout -b ibaranga-master master
git pull https://github.com/ibaranga/openapi-generator.git master
```

ibaranga commented:
@wing328, @davidkarlsen, the existing `spring-http-interface` generator is using internally the `HttpServiceProxyFactory#builder(HttpClientAdapter)` factory method.

This `HttpClientAdapter` interface / `HttpServiceProxyFactory#builder(HttpClientAdapter)` factory method:

- were **deprecated** in Spring Boot 3.2 in favor of `HttpServiceProxyFactory#builderFor(HttpExchangeAdapter)`.
- were **removed** in Spring Boot 3.4
- can only be used with WebClient, since there is no spring-provided HttpClientAdapter for RestClient / RestTemplate, unlike the `HttpExchangeAdapter` replacement

Considering this deprecation and the complexity of maintaining configurations and samples for 3 different client 
implementations, I’ve introduced the possibility to generate a `HttpInterfacesAbstractConfigurator` which depends on a provided `HttpServiceProxyFactory` instance, rather than specific Spring client implementations. 

This can be enabled by setting a new property called `useHttpServiceProxyFactoryInterfacesConfigurator` to `"true"`;

To ensure backward compatibility with Spring Boot 3.1 and flexibility for Spring Boot 3.2+:

- When `useHttpServiceProxyFactoryInterfacesConfigurator` **is missing or false**
  - keep using the Spring Boot 3.1 API (`HttpServiceProxyFactory#builder(WebClientAdapter(webClient))`) in generated `HttpInterfacesAbstractConfigurator`  classes, based on a provided `WebClient` instance
  - this will remain functional until Spring Boot 3.4.

- When `useHttpServiceProxyFactoryInterfacesConfigurator` **is true**,
  - generate `HttpInterfacesAbstractConfigurator` classes that depend on a `HttpServiceProxyFactory` instance, allowing greater flexibility in selecting the underlying HTTP client



The changes on custom `HttpInterfacesAbstractConfigurator` extensions will be minimal, for example  

```diff
@Configuration
public class MyConfiguration extends HttpInterfacesAbstractConfigurator {
   // ...
}

// WebClient example
@Bean
public  MyConfiguration myConfiguration(WebClient webClient) {
---      return new MyConfiguration(webClient)
+++      return new MyConfiguration(HttpServiceProxyFactory.builderFor(WebClientAdapter.create(webClient)));
}

// RestTemplate example
@Bean
public MyConfiguration myConfiguration(RestTemplate restTemplate) {
---     return new MyConfiguration(restTemplate)
+++     return new MyConfiguration(HttpServiceProxyFactory.builderFor(RestTemplateAdapter.create(restTemplate)));
}

```

DuncanCasteleyn commented:
This is blocking our upgrade to spring 3.4.x. We could really use this pull request.

DuncanCasteleyn commented:
@wing328 Could you please confirm to us that [7.12.0](https://github.com/OpenAPITools/openapi-generator/milestone/65) will contain this pull request we are awaiting this change at out company for a spring update.

We need this information so we can plan ahead.

martin-mfg commented on bin/configs/spring-http-interface-useHttpServiceProxyFactoryInterfacesConfiguration.yaml:
```diff
@@ -0,0 +1,21 @@
+  useBeanValidation: "true"
+  performBeanValidation: "true"
+  useHttpServiceProxyFactoryInterfacesConfigurator: "true"
+  # needed until updating the default parentVersion (3.1.3) to 3.2+
```
Is this still true after removing the two imports mentioned in my other comment?

ibaranga replied:
Good point, removed the parent overrides!

martin-mfg commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:
```diff
@@ -268,6 +271,10 @@ public SpringCodegen() {
         cliOptions.add(new CliOption(RESOURCE_FOLDER, RESOURCE_FOLDER_DESC).defaultValue(this.getResourceFolder()));
 
+        cliOptions.add(CliOption.newString(USE_HTTP_SERVICE_PROXY_FACTORY_INTERFACES_CONFIGURATOR,
+            "Generate HttpInterfacesAbstractConfigurator based on an HttpServiceProxyFactory instance (as opposed to a WebClient instance, when disabled) for generating Spring HTTP interfaces. Requires spring-web 6.1+.")
```
>Requires spring-web 6.1+.

Is this still true after removing the two imports mentioned in my other comment?

martin-mfg replied:
The mention of spring-web 6.1+ is still there.

martin-mfg reviewed (changes requested):
Hi @ibaranga, thanks for the PR! Could you please resolve the merge conflicts and have a look at the review comments below?
Also, please provide an example configuration and input specification for which the `@Parameter` problem occurs.

ibaranga commented on modules/openapi-generator/src/main/resources/JavaSpring/libraries/spring-http-interface/httpServiceProxyFactoryInterfacesConfigurator.mustache:
```diff
@@ -0,0 +1,37 @@
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.web.client.RestClient;
+import org.springframework.web.client.support.RestClientAdapter;
```
Fixed!

ibaranga commented:
> Hi @ibaranga, thanks for the PR! Could you please resolve the merge conflicts and have a look at the review comments below? Also, please provide an example configuration and input specification for which the `@Parameter` problem occurs.

Hi @martin-mfg, I've addressed the PR comments.

Regarding the @Parameter problem: I can reproduce it only when generating with the Gradle plugin (but not via the CLI).
I’ve prepared an example in this sample repo: https://github.com/ibaranga/openapi-generator-spring-http-interface-example

martin-mfg commented:
I analyzed your gradle/`@Parameter` issue and found the problem: Instead of specifying `configOptions = [library : 'spring-http-interface']` you should specify `library = 'spring-http-interface'`. This is the intended way, and this generates what you expect. It would be nice if you could add a warning in the gradle plugin for situations where users make this (understandable) mistake. In case you don't have time for that, please at least revert the addition of paramDoc.mustache.

martin-mfg reviewed (changes requested):
I analyzed your gradle/`@Parameter` issue and found the problem: Instead of specifying `configOptions = [library : 'spring-http-interface']` you should specify `library = 'spring-http-interface'`. This is the intended way, and this generates what you expect. It would be nice if you could add a warning in the gradle plugin for situations where users make this (understandable) mistake. In case you don't have time for that, please at least revert the addition of paramDoc.mustache.

ibaranga commented:
> I analyzed your gradle/`@Parameter` issue and found the problem: Instead of specifying `configOptions = [library : 'spring-http-interface']` you should specify `library = 'spring-http-interface'`. This is the intended way, and this generates what you expect. It would be nice if you could add a warning in the [gradle plugin](https://github.com/OpenAPITools/openapi-generator/blob/master/modules/openapi-generator-gradle-plugin/src/main/kotlin/org/openapitools/generator/gradle/plugin/tasks/GenerateTask.kt) for situations where users make this (understandable) mistake. In case you don't have time for that, please at least revert the addition of paramDoc.mustache.

Thanks for clarifying - for now I just reverted the addition of paramDoc.mustache.

Bragolgirith commented:
I wouldn't want to steal the attribution from the original PR author, so (for now) I've now created a [PR](https://github.com/ibaranga/openapi-generator/pull/1) to @ibaranga's fork with the following changes:

* Merged from `master` and resolved conflicts
* Added a check to fail early if a Spring Boot version other than 3 or 4 is used

@ibaranga Would you be willing to take a look and - if at all possible - merge it soon-ish? :)
